features/armv8_stack_overflow_guards/README.md

```diff
@@ -0,0 +1,194 @@
+ UsageFault_Handler:
+ /*
+	 * We need special handling for the UsageFault_Handler with the use of the MSPLIM register set.  First we check to
+	 * see if the MSPLIM or the PSPLIM registers equal the msp or psp and set the stack pointer mask respectively.
+	 * If the MSPLIM register triggered the usagefault, then we must move the MSPLIM back below the padding to allow for the
+	 * UsageFault_HandlerC to push to the stack if needed.  If the MSPLIM is not moved, then a push to the stack can cause
+	 * nested UsageFaults and circular exceptions
+	 */
+		       ldr		r2,=UsageFault_HandlerC
+		       mrs		r1, msp
+		       mrs		r0, MSPLIM
+		       cmp		r0, r1
+		       beq		UsageFault_with_MSP_Overflow
+		       mrs		r1, psp
+		       mrs		r0, PSPLIM
+		       cmp		r0, r1
+		       beq 		UsageFault_with_PSP_Overflow
+		       mov		r0, #0
+		       bx		r1
+
+ UsageFault_with_PSP_Overflow:
+		       mov 		r0, #2
+		       bx		r2
+
+ UsageFault_with_MSP_Overflow:
+		       ldr 		r1, =__StackLimit
+		       msr		MSPLIM, r1
+		       mov		r0, #1
+		       bx		r2
+ #endif
+
+  ```
+
+- **FreeRTOS Modifications - for PSP protection**
+We need to make two small changes to FreeRTOS to allow for switching out the PSP.  Navigate to sdk/FreeRTOS/tasks.c
+
+Add the following line of code above *vTaskSwitchContext*:  
+
+
+```c
+#if dg_configARMV8_USE_STACK_GUARDS == 1
+
+void vTaskSwitchStackGuard(void)
+{
+        volatile uint32_t end_of_stack_val = (uint32_t)pxCurrentTCB->pxStack;
+     __set_PSPLIM( end_of_stack_val);
+}
+
+#endif //#if dg_configARMV8_USE_STACK_GUARDS == 1
+
+```
+Next, modify the xPortPendSVHandler to switch in the PSPLIM guards, in sdk/FreeRTOS/protable/GCC/DA1469x/port.c:
+
+```diff
+
+	"	bl vTaskSwitchContext				\n"
++#if dg_configARMV8_USE_STACK_GUARDS == 1
++ "     bl vTaskSwitchStackGuard                        \n"
++ #endif
+	"	mov r0, #0
+
+```
+
+- **Fault Handler Modification**
+Last, modify the UsageFault_Handler in peripherals/hw_hard_fault.c to have a weak attribute:
+
+```c
+__WEAK void UsageFault_Handler(void)
+```
+
+
+
+
+## Running the example
+
+### Initial Setup
+
+- Clone the Source code from Github.
+- Import the project into your workspace.
+- Connect the target device to your host PC.
+- Compile the code (either in Release or Debug Mode)  and load it onto the chip.
+- Place a breakpoint in the UsageFault_Handler in overflow_task.c
+- Short Press the K1 Button on the Pro Kit to trigger an MSP or PSP overflow.   
+
+
+### Modifying the example to generate other issues.
+
+The example uses a recursive function call to illustrate the limitations of FreeRTOS and software solutions.  We can modify the example with switches:
+
+  - in overflow_task.c, there is a define - TOGGLE_MSP_OVERFLOW. (0) Will create a PSP overflow, and (1) will create an MSP overflow.
+  - In custom_config_xxx.h, you can #undef dg_configARMV8_USE_STACK_GUARDS, and this will illustrate what happens when we don't use the hardware guards.
```

Comment:
An image or some description of what happens when dg_configARMV8_USE_STACK_GUARDS is undefined would be useful